File: ts-sdk/src/_internal/discovery.ts
// Copyright (c) Aptos Labs
// SPDX-License-Identifier: Apache-2.0

import { AccountAddress, Deserializer } from "@aptos-labs/ts-sdk";
import { Element as GroupElement } from "../group";
import * as pke from "../pke";
import { State as NetworkState } from "../network";

/** Human-readable projection of `DiscoveryViewV0` (all crypto material hex-encoded). */
export interface DiscoveryReadableV0 {
    epoch: number;
    epochChanging: boolean;
    threshold: number;
    epochStartTimeMicros: string;
    epochDurationMicros: string;
    nodes: { address: string; endpoint: string | null; pkeEncKey: string | null }[];
    keypairs: {
        keypairId: string;
        currentSession: string;
        scheme: number;
        note: string;
        masterPublicKey: string | null;
        basePoint: string | null;
        sharePks: string[];
    }[];
}

/** The base point, master public key, and per-holder share PKs of a secret's session. */
export interface SessionPks {
    basePoint: GroupElement;
    sharePks: GroupElement[];
    /** Master public key. Present for DKG sessions (used for the encryption key); may be omitted for DKR. */
    resultPk?: GroupElement;
}

/**
 * Typed, read-only view over the ACE on-chain state the SDK needs. Two implementations:
 * a fullnode-backed one (`FullnodeChainReader` in common.ts, one view call per method) and the
 * discovery-backed `DiscoveryChainReader` below (all methods answered from one aggregated
 * snapshot). `getChainReader` (common.ts) picks between them per `AceDeployment`.
 */
export interface ChainReader {
    networkState(): Promise<NetworkState>;
    /** `isDkg` selects the session module for the fullnode path; discovery ignores it (uniform). */
    session(addr: string, isDkg: boolean): Promise<SessionPks>;
    workerEndpoint(addr: string): Promise<string>;
    workerEncKey(addr: string): Promise<pke.EncryptionKey>;
}

/** Decoded `ace::network::DiscoveryViewV0` (see contracts/network/sources/network.move). */
export class DiscoveryViewV0 {
    constructor(
        readonly state: NetworkState,
        /** node address (long form) -> { endpoint?, encKey }. */
        readonly nodes: Map<string, { endpoint?: string; encKey: pke.EncryptionKey }>,
        /** session address (long form) -> SessionPks. */
        readonly sessions: Map<string, SessionPks>,
    ) {}

    /**
     * A plain, JSON-serializable, human-readable projection — all crypto material as hex, epoch
     * timers as decimal strings (bigint isn't JSON-serializable). Intended for the discovery
     * service's `/json` debug endpoint and for eyeballing; the SDK itself decodes the BCS directly.
     */
    toReadable(): DiscoveryReadableV0 {
        const nodes = this.state.curNodes.map((addr) => {
            const key = addr.toStringLong();
            const n = this.nodes.get(key);
            return { address: key, endpoint: n?.endpoint ?? null, pkeEncKey: n ? n.encKey.toHex() : null };
        });
        const keypairs = this.state.secrets.map((s) => {
            const keypairId = s.keypairId.toStringLong();
            const currentSession = s.currentSession.toStringLong();
            const cur = this.sessions.get(currentSession);
            const origin = this.sessions.get(keypairId); // keypairId is always the origin DKG session
            return {
                keypairId,
                currentSession,
                scheme: s.scheme,
                note: s.note,
                masterPublicKey: origin?.resultPk?.toHex() ?? null,
                basePoint: cur?.basePoint.toHex() ?? null,
                sharePks: cur ? cur.sharePks.map((p) => p.toHex()) : [],
            };
        });
        return {
            epoch: this.state.epoch,
            epochChanging: this.state.isEpochChanging(),
            threshold: this.state.curThreshold,
            epochStartTimeMicros: this.state.epochStartTimeMicros.toString(),
            epochDurationMicros: this.state.epochDurationMicros.toString(),
            nodes,
            keypairs,
        };
    }

    static fromBytes(bytes: Uint8Array): DiscoveryViewV0 {
        const d = new Deserializer(bytes);

        // state_view: an inline StateViewV0, byte-identical to state_view_v0_bcs() — reuse State's decoder.
        const state = NetworkState.deserialize(d).unwrapOrThrow("DiscoveryViewV0: parse state_view");

        const nodes = new Map<string, { endpoint?: string; encKey: pke.EncryptionKey }>();
        const nodesLen = d.deserializeUleb128AsU32();
        for (let i = 0; i < nodesLen; i++) {
            const addr = AccountAddress.deserialize(d).toStringLong();
            const tag = d.deserializeU8(); // Option<String>
            let endpoint: string | undefined;
            if (tag === 1) {
                endpoint = d.deserializeStr();
            } else if (tag !== 0) {
                throw new Error(`DiscoveryViewV0: node endpoint option tag must be 0 or 1, got ${tag}`);
            }
            const encKey = pke.EncryptionKey.deserialize(d).unwrapOrThrow(`DiscoveryViewV0: parse enc_key for ${addr}`);
            nodes.set(addr, { endpoint, encKey });
        }

        const sessions = new Map<string, SessionPks>();
        const sessionsLen = d.deserializeUleb128AsU32();
        for (let i = 0; i < sessionsLen; i++) {
            const addr = AccountAddress.deserialize(d).toStringLong();
            const basePoint = GroupElement.deserialize(d).unwrapOrThrow(`DiscoveryViewV0: parse base_point for ${addr}`);
            const resultPk = GroupElement.deserialize(d).unwrapOrThrow(`DiscoveryViewV0: parse result_pk for ${addr}`);
            const sharePksLen = d.deserializeUleb128AsU32();
            const sharePks: GroupElement[] = [];
            for (let j = 0; j < sharePksLen; j++) {
                sharePks.push(GroupElement.deserialize(d).unwrapOrThrow(`DiscoveryViewV0: parse share_pks[${j}] for ${addr}`));
            }
            sessions.set(addr, { basePoint, sharePks, resultPk });
        }

        if (d.remaining() !== 0) throw new Error("DiscoveryViewV0: trailing bytes");
        return new DiscoveryViewV0(state, nodes, sessions);
    }
}

/** GET the aggregated snapshot from the discovery service and decode it. */
async function fetchDiscoveryView(discoveryUrl: string): Promise<DiscoveryViewV0> {
    const resp = await fetch(discoveryUrl);
    if (!resp.ok) {
        throw new Error(`ACE discovery: GET ${discoveryUrl} -> HTTP ${resp.status}`);
    }
    let hex = (await resp.text()).trim();
    // The server serves the plain 0x hex body; also tolerate a { discoveryViewV0Bcs } JSON envelope.
    if (hex.startsWith("{")) {
        const obj = JSON.parse(hex) as { discoveryViewV0Bcs?: string };
        if (!obj.discoveryViewV0Bcs) throw new Error("ACE discovery: response JSON missing 'discoveryViewV0Bcs'");
        hex = obj.discoveryViewV0Bcs;
    }
    hex = hex.replace(/^0x/, "");
    const bytes = Uint8Array.from(hex.match(/.{1,2}/g)?.map((b) => parseInt(b, 16)) ?? []);
    return DiscoveryViewV0.fromBytes(bytes);
}

function addrKey(addr: string): string {
    return AccountAddress.fromString(addr).toStringLong();
}

/** Answers the typed reads from a single lazily-fetched discovery snapshot. */
export class DiscoveryChainReader implements ChainReader {
    private snapshot?: Promise<DiscoveryViewV0>;

    constructor(private readonly discoveryUrl: string) {}

    private view0(): Promise<DiscoveryViewV0> {
        // Fetch once per reader instance (i.e. once per operation phase). No cross-op cache: a
        // fresh reader is created per call site, so state stays current across epoch changes.
        if (this.snapshot === undefined) this.snapshot = fetchDiscoveryView(this.discoveryUrl);
        return this.snapshot;
    }

    async networkState(): Promise<NetworkState> {
        return (await this.view0()).state;
    }

    async session(addr: string): Promise<SessionPks> {
        const s = (await this.view0()).sessions.get(addrKey(addr));
        if (s === undefined) throw new Error(`ACE discovery: session ${addrKey(addr)} not in snapshot`);
        return s;
    }

    async workerEndpoint(addr: string): Promise<string> {
        const n = (await this.view0()).nodes.get(addrKey(addr));
        if (n === undefined) throw new Error(`ACE discovery: node ${addrKey(addr)} not in snapshot`);
        if (n.endpoint === undefined) throw new Error(`ACE discovery: node ${addrKey(addr)} has no registered endpoint`);
        return n.endpoint;
    }

    async workerEncKey(addr: string): Promise<pke.EncryptionKey> {
        const n = (await this.view0()).nodes.get(addrKey(addr));
        if (n === undefined) throw new Error(`ACE discovery: node ${addrKey(addr)} not in snapshot`);
        return n.encKey;
    }
}

File: python-sdk/src/ace_sdk/_internal/discovery.py
# Copyright (c) Aptos Labs
# SPDX-License-Identifier: Apache-2.0
"""Discovery snapshot decoder, mirroring ts-sdk/src/_internal/discovery.ts."""

from __future__ import annotations

import json
from dataclasses import dataclass
from urllib.request import urlopen

from aptos_sdk.account_address import AccountAddress

from ace_sdk import pke
from ace_sdk.bcs import Deserializer, deserialize_account_address
from ace_sdk.group import Element
from ace_sdk.network import State


@dataclass(frozen=True)
class SessionPks:
    base_point: Element
    share_pks: list[Element]
    result_pk: Element | None = None


def _addr_key(addr: str | AccountAddress) -> str:
    if isinstance(addr, str):
        account_addr = AccountAddress.from_str(addr)
    else:
        account_addr = addr
    return "0x" + account_addr.address.hex()


class DiscoveryViewV0:
    def __init__(
        self,
        state: State,
        nodes: dict[str, dict[str, str | pke.EncryptionKey]],
        sessions: dict[str, SessionPks],
    ) -> None:
        self.state = state
        self.nodes = nodes
        self.sessions = sessions

    def to_readable(self) -> dict:
        nodes = []
        for addr in self.state.cur_nodes:
            key = _addr_key(addr)
            node = self.nodes.get(key)
            nodes.append(
                {
                    "address": key,
                    "endpoint": node.get("endpoint") if node else None,
                    "pkeEncKey": node["enc_key"].to_hex() if node else None,
                }
            )

        keypairs = []
        for secret in self.state.secrets:
            keypair_id = _addr_key(secret.keypair_id)
            current_session = _addr_key(secret.current_session)
            cur = self.sessions.get(current_session)
            origin = self.sessions.get(keypair_id)
            keypairs.append(
                {
                    "keypairId": keypair_id,
                    "currentSession": current_session,
                    "scheme": secret.scheme,
                    "note": secret.note,
                    "masterPublicKey": origin.result_pk.to_hex()
                    if origin and origin.result_pk
                    else None,
                    "basePoint": cur.base_point.to_hex() if cur else None,
                    "sharePks": [share_pk.to_hex() for share_pk in cur.share_pks]
                    if cur
                    else [],
                }
            )

        return {
            "epoch": self.state.epoch,
            "epochChanging": self.state.is_epoch_changing(),
            "threshold": self.state.cur_threshold,
            "epochStartTimeMicros": str(self.state.epoch_start_time_micros),
            "epochDurationMicros": str(self.state.epoch_duration_micros),
            "nodes": nodes,
            "keypairs": keypairs,
        }

    @staticmethod
    def from_bytes(data: bytes) -> "DiscoveryViewV0":
        deserializer = Deserializer(data)
        state = State.deserialize(deserializer).unwrap_or_throw(
            ValueError("DiscoveryViewV0: parse state_view")
        )

        nodes: dict[str, dict[str, str | pke.EncryptionKey]] = {}
        nodes_len = deserializer.deserialize_uleb128_as_u32()
        for _ in range(nodes_len):
            addr = _addr_key(deserialize_account_address(deserializer))
            tag = deserializer.deserialize_u8()
            endpoint: str | None = None
            if tag == 1:
                endpoint = deserializer.deserialize_str()
            elif tag != 0:
                raise ValueError(
                    f"DiscoveryViewV0: node endpoint option tag must be 0 or 1, got {tag}"
                )
            enc_key = pke.EncryptionKey.deserialize(deserializer).unwrap_or_throw(
                ValueError(f"DiscoveryViewV0: parse enc_key for {addr}")
            )
            node: dict[str, str | pke.EncryptionKey] = {"enc_key": enc_key}
            if endpoint is not None:
                node["endpoint"] = endpoint
            nodes[addr] = node

        sessions: dict[str, SessionPks] = {}
        sessions_len = deserializer.deserialize_uleb128_as_u32()
        for _ in range(sessions_len):
            addr = _addr_key(deserialize_account_address(deserializer))
            base_point = Element.deserialize(deserializer).unwrap_or_throw(
                ValueError(f"DiscoveryViewV0: parse base_point for {addr}")
            )
            result_pk = Element.deserialize(deserializer).unwrap_or_throw(
                ValueError(f"DiscoveryViewV0: parse result_pk for {addr}")
            )
            share_pks_len = deserializer.deserialize_uleb128_as_u32()
            share_pks = []
            for j in range(share_pks_len):
                share_pks.append(
                    Element.deserialize(deserializer).unwrap_or_throw(
                        ValueError(f"DiscoveryViewV0: parse share_pks[{j}] for {addr}")
                    )
                )
            sessions[addr] = SessionPks(base_point=base_point, share_pks=share_pks, result_pk=result_pk)

        if deserializer.remaining() != 0:
            raise ValueError("DiscoveryViewV0: trailing bytes")
        return DiscoveryViewV0(state, nodes, sessions)


def _hex_to_bytes(hex_str: str) -> bytes:
    h = hex_str.strip()
    if h.startswith("0x") or h.startswith("0X"):
        h = h[2:]
    return bytes.fromhex(h)


def fetch_discovery_view(discovery_url: str) -> DiscoveryViewV0:
    url = discovery_url.rstrip("/") + "/bcs"
    with urlopen(url) as response:
        status = getattr(response, "status", 200)
        if status < 200 or status >= 300:
            raise RuntimeError(f"ACE discovery: GET {url} -> HTTP {status}")
        body = response.read().decode("utf-8").strip()
    if body.startswith("{"):
        obj = json.loads(body)
        body = obj.get("discoveryViewV0Bcs")
        if not body:
            raise ValueError("ACE discovery: response JSON missing 'discoveryViewV0Bcs'")
    return DiscoveryViewV0.from_bytes(_hex_to_bytes(body))


class DiscoveryChainReader:
    def __init__(self, discovery_url: str) -> None:
        self.discovery_url = discovery_url
        self._snapshot: DiscoveryViewV0 | None = None

    def _view0(self) -> DiscoveryViewV0:
        if self._snapshot is None:
            self._snapshot = fetch_discovery_view(self.discovery_url)
        return self._snapshot

    def network_state(self) -> State:
        return self._view0().state

    def session(self, addr: str, is_dkg: bool | None = None) -> SessionPks:
        del is_dkg
        key = _addr_key(addr)
        session = self._view0().sessions.get(key)
        if session is None:
            raise ValueError(f"ACE discovery: session {key} not in snapshot")
        return session

    def worker_endpoint(self, addr: str) -> str:
        key = _addr_key(addr)
        node = self._view0().nodes.get(key)
        if node is None:
            raise ValueError(f"ACE discovery: node {key} not in snapshot")
        endpoint = node.get("endpoint")
        if endpoint is None:
            raise ValueError(f"ACE discovery: node {key} has no registered endpoint")
        return endpoint

    def worker_enc_key(self, addr: str) -> pke.EncryptionKey:
        key = _addr_key(addr)
        node = self._view0().nodes.get(key)
        if node is None:
            raise ValueError(f"ACE discovery: node {key} not in snapshot")
        return node["enc_key"]  # type: ignore[return-value]


__all__ = [
    "DiscoveryChainReader",
    "DiscoveryViewV0",
    "SessionPks",
    "fetch_discovery_view",
]
